tools/dashboard/generate_dashboard.py

```diff
@@ -381,15 +466,23 @@ def make_plot_image(fn, title, ylabel, curves, max_age, full_archive):
 
 
 # ======================================================================================
-def send_notification(report, last_ok, log, name, s):
-    idx_end = log.index[report["git-sha"]] if (report["git-sha"]) else 0
+def send_notification(
+    report: Report,
+    last_ok: Optional[GitSha],
+    log: GitLog,
+    name: str,
+    s: str,
+    send_emails: bool,
+) -> None:
+
+    idx_end = log.index_by_sha[report.sha] if report.sha else 0
     if not last_ok:
         return  # we don't know when this started
-    idx_last_ok = log.index[last_ok]
+    idx_last_ok = log.index_by_sha[last_ok]
     if idx_end == idx_last_ok:
         return  # probably a flapping tester
-    emails = set([log[i]["author-email"] for i in range(idx_end, idx_last_ok)])
-    emails = [e for e in emails if "noreply" not in e]
+    emails_raw = set([log.commits[i].author_email for i in range(idx_end, idx_last_ok)])
```

Comment:
no need to create a list first and you can also drop the explicit index:
`set(c.author_email for c in log.commits[idx_end:idx_last_ok])`

Reply:
Good idea. Thanks for the thorough review! I learned a lot :-)